File: mandi-direct/frontend/src/types/order.ts
import { OrderLogistics } from "./logistics";

export type OrderStatus =
  | "PENDING"
  | "ACCEPTED"
  | "PREPARING"
  | "READY_FOR_PICKUP"
  | "PICKED_UP"
  | "OUT_FOR_DELIVERY"
  | "DELIVERED"
  | "CANCELLED"
  | "REJECTED";

export type PaymentStatus = "PENDING" | "PAID" | "FAILED" | "REFUNDED";

export interface BuyerAddress {
  id: string;
  buyer_user_id: string;
  full_name: string;
  phone: string;
  address_line1: string;
  address_line2?: string | null;
  village?: string | null;
  mandal?: string | null;
  district: string;
  state: string;
  pincode: string;
  landmark?: string | null;
  is_default: boolean;
  created_at: string;
  updated_at: string;
}

export interface BuyerAddressCreate {
  full_name: string;
  phone: string;
  address_line1: string;
  address_line2?: string;
  village?: string;
  mandal?: string;
  district: string;
  state: string;
  pincode: string;
  landmark?: string;
  is_default?: boolean;
}

export interface CartProductInfo {
  id: string;
  product_name: string;
  category: string;
  variety?: string | null;
  expected_price: number;
  price_unit: string;
  quantity_unit: string;
  available_quantity: number;
  minimum_order_quantity: number;
  primary_image_url?: string | null;
  status: string;
  farmer_name: string;
  farmer_location: string;
}

export interface CartItem {
  id: string;
  cart_id: string;
  produce_listing_id: string;
  quantity: number;
  product: CartProductInfo;
  subtotal: number;
  created_at: string;
  updated_at: string;
}

export interface CartResponse {
  id: string;
  buyer_user_id: string;
  items: CartItem[];
  item_count: number;
  subtotal: number;
  delivery_fee: number;
  total_amount: number;
}

export interface OrderItem {
  id: string;
  order_id: string;
  produce_listing_id: string;
  farmer_profile_id: string;
  product_name: string;
  quantity: number;
  quantity_unit: string;
  unit_price: number;
  subtotal: number;
  primary_image_url?: string | null;
  farmer_name?: string | null;
}

export interface OrderStatusHistory {
  id: string;
  order_id: string;
  previous_status?: string | null;
  new_status: string;
  changed_by: string;
  actor_name?: string | null;
  reason?: string | null;
  created_at: string;
}

export interface BuyerOrderSummary {
  id: string;
  order_number: string;
  status: OrderStatus;
  payment_status: PaymentStatus;
  subtotal: number;
  delivery_fee: number;
  total_amount: number;
  items_count: number;
  first_product_name: string;
  first_product_image?: string | null;
  created_at: string;
}

export interface BuyerOrderDetail {
  id: string;
  order_number: string;
  status: OrderStatus;
  payment_status: PaymentStatus;
  subtotal: number;
  delivery_fee: number;
  total_amount: number;
  delivery_address_snapshot: BuyerAddress;
  buyer_notes?: string | null;
  items: OrderItem[];
  status_history: OrderStatusHistory[];
  logistics?: OrderLogistics | null;
  created_at: string;
  updated_at: string;
}

export interface FarmerOrderItem {
  id: string;
  produce_listing_id: string;
  product_name: string;
  variety?: string | null;
  quality_grade?: string | null;
  quantity: number;
  quantity_unit: string;
  unit_price: number;
  subtotal: number;
}

export interface FarmerOrderSummary {
  id: string;
  order_number: string;
  status: OrderStatus;
  created_at: string;
  items_count: number;
  product_names?: string[];
  farmer_subtotal: number;
  buyer_name: string;
  delivery_district: string;
  delivery_state: string;
  collection_point_name?: string | null;
}

export interface FarmerOrderDetail {
  id: string;
  order_number: string;
  status: OrderStatus;
  created_at: string;
  buyer_name: string;
  buyer_phone?: string | null;
  delivery_address: BuyerAddress;
  buyer_notes?: string | null;
  items: FarmerOrderItem[];
  farmer_subtotal: number;
  status_history: OrderStatusHistory[];
  logistics?: OrderLogistics | null;
}

export interface FarmerOrderStats {
  pending: number;
  accepted: number;
  preparing: number;
  ready_for_pickup: number;
  picked_up: number;
  out_for_delivery: number;
  delivered: number;
  rejected: number;
  cancelled: number;
  total_requiring_action: number;
}

File: mandi-direct/backend/app/schemas/order.py
from datetime import datetime
from decimal import Decimal
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field
from app.models.enums import OrderStatus, PaymentStatus
from app.schemas.logistics import OrderLogisticsResponse


class OrderCreateRequest(BaseModel):
    delivery_address_id: str = Field(..., description="ID of saved delivery address belonging to the authenticated buyer")
    buyer_notes: Optional[str] = Field(None, max_length=1000, description="Optional delivery or packing notes")


class OrderStatusUpdateRequest(BaseModel):
    status: OrderStatus = Field(..., description="Target order status")
    reason: Optional[str] = Field(None, max_length=500, description="Reason for status transition or rejection")


class OrderCancelRequest(BaseModel):
    reason: Optional[str] = Field(None, max_length=500, description="Cancellation reason")


class OrderItemResponse(BaseModel):
    id: str
    order_id: str
    produce_listing_id: str
    farmer_profile_id: str
    product_name: str
    quantity: Decimal
    quantity_unit: str
    unit_price: Decimal
    subtotal: Decimal
    primary_image_url: Optional[str] = None
    farmer_name: Optional[str] = None

    class Config:
        from_attributes = True


class OrderStatusHistoryResponse(BaseModel):
    id: str
    order_id: str
    previous_status: Optional[str] = None
    new_status: str
    changed_by: str
    actor_name: Optional[str] = None
    reason: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class BuyerOrderSummaryResponse(BaseModel):
    id: str
    order_number: str
    status: str
    payment_status: str
    subtotal: Decimal
    delivery_fee: Decimal
    total_amount: Decimal
    items_count: int
    first_product_name: str
    first_product_image: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class BuyerOrderDetailResponse(BaseModel):
    id: str
    order_number: str
    status: str
    payment_status: str
    subtotal: Decimal
    delivery_fee: Decimal
    total_amount: Decimal
    delivery_address_snapshot: Dict[str, Any]
    buyer_notes: Optional[str] = None
    items: List[OrderItemResponse]
    status_history: List[OrderStatusHistoryResponse]
    logistics: Optional[OrderLogisticsResponse] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class FarmerOrderItemResponse(BaseModel):
    id: str
    produce_listing_id: str
    product_name: str
    variety: Optional[str] = None
    quality_grade: Optional[str] = None
    quantity: Decimal
    quantity_unit: str
    unit_price: Decimal
    subtotal: Decimal

    class Config:
        from_attributes = True


class FarmerOrderSummaryResponse(BaseModel):
    id: str
    order_number: str
    status: str
    created_at: datetime
    items_count: int
    product_names: Optional[List[str]] = []
    farmer_subtotal: Decimal
    buyer_name: str
    delivery_district: str
    delivery_state: str
    collection_point_name: Optional[str] = None

    class Config:
        from_attributes = True


class FarmerOrderDetailResponse(BaseModel):
    id: str
    order_number: str
    status: str
    created_at: datetime
    buyer_name: str
    buyer_phone: Optional[str] = None
    delivery_address: Dict[str, Any]
    buyer_notes: Optional[str] = None
    items: List[FarmerOrderItemResponse]
    farmer_subtotal: Decimal
    status_history: List[OrderStatusHistoryResponse]
    logistics: Optional[OrderLogisticsResponse] = None

    class Config:
        from_attributes = True


class FarmerOrderStatsResponse(BaseModel):
    pending: int = 0
    accepted: int = 0
    preparing: int = 0
    ready_for_pickup: int = 0
    picked_up: int = 0
    out_for_delivery: int = 0
    delivered: int = 0
    rejected: int = 0
    cancelled: int = 0
    total_requiring_action: int = 0
